### DISCUSSION TITLE
What is the benefit of Resolver

### DISCUSSION THREAD

**Discussion by lmunck** [peterfriese/MakeItSo#57] [Q&A] [answered]
What is the benefit of dependency injection with Resolver instead of simply calling that class directly?

in both cases the class can be part of the state, so is it because the client class only loads part of the service class to save memory or is there something else going on?

**Comment by peterfriese** [comment, marked as answer]
Hi Anders,

thanks for asking!

The `AuthenticationService` is used in a couple of places in the app:
- In the Settings screen (via [`SettingsViewModel`](https://github.com/peterfriese/MakeItSo/blob/master/code/frontend/iOS/MakeItSo/MakeItSo/ViewModels/SettingsViewModel.swift))
- In the main app, to trigger sign-in at app start-up ([`MakeItSoApp `](https://github.com/peterfriese/MakeItSo/blob/master/code/frontend/iOS/MakeItSo/MakeItSo/ViewModels/SettingsViewModel.swift))
- And in the [`TaskRepository`](https://github.com/peterfriese/MakeItSo/blob/master/code/frontend/iOS/MakeItSo/MakeItSo/Repositories/TaskRepository.swift)

When I started creating the app, I was hoping to be able to use `@EnvironmentObject` to inject the service where I needed it, but unfortunately it is not possible to fetch the environment when you're _inside_ an `ObservableObject`.

This is why I decided to use a lightweight dependency injection container. Resolver strikes a nice balance of being extremely lightweight yet powerful, and to this day I use it in many of my apps.

Now, obviously you could implement this differently - for example, you could configure the `ObervableObject` once the view has been initialised, or you could just instantiate multiple instances of the `AuthenticationService`. Still, I think this implementation is rather clean and allows for some added flexibility (and makes the code more testable).

Hope this helped understand why I chose this path. Feel free to ask if anything is unclear.

Cheers,
Peter
